Add Physics for reaction network-based problems

GiudGiud opened the pull request:
~~Not finished. Early PR to get feedback on class names and input~~

The new input syntax:

```
[Physics]
  [AqueousReactionsEquilibrium]
    [water]
      block = 'water'
      ...
      reactions = 'A + 2*B- = C     [K=3]
                  C + 3.5*D2 = E    [K=2]
```
The `[]` could welcome other things like decay constants, thresholds etc, depending on the type of reaction at play (plasma kinetics more complicated than others for example)

See #31453 for parser
See #31454 for Physics syntax

I am happy to do the solid kinetics Physics once this is accepted.
Once we have both, I think we should deprecate the `[ReactionNetwork]` syntax. I definitely used a lot of that code (just not the parser) so I would say we are very much building on prior work here.

GiudGiud committed:
Fix logic when multiple variables exist for diffusion physics problem

GiudGiud committed:
Add a util to parse the reaction network using the parsed expression grammar (PEG)
- should be easier to maintain than the PCRE parser
refs #31453

GiudGiud committed:
Add a base class for reaction kinetics
refs #31454

GiudGiud committed:
Add a derived class for aqueous reaction kinetics
refs #31454

GiudGiud committed:
Adapt to new reaction kinetics parsing
refs #31453

GiudGiud committed:
Sort an advanced tagging parameter in the Advanced param group

GiudGiud committed:
Add more parsing, move over kernel creation code
refs #31453

GiudGiud committed:
Improve messages on missed parameters for Physics

GiudGiud committed:
Add a stringify overload for reactions
refs #31453

GiudGiud committed:
Add some utils to work with the Reaction struct
refs #31453

GiudGiud committed:
Add parsing of the reactions for aqueous reaction
refs #31454

GiudGiud committed:
Add tests for physics implementation of aqueous reaction kinetics
refs #31454

GiudGiud committed:
Fix the parser:
- string_view vs string
- add asserts
- add capability for the minus sign
- fix logic for charge and state
refs #31453

GiudGiud committed:
Register new Physics syntax
refs #31454

GiudGiud committed:
Use missed parameter report with variable name
Keep the reaction output in its input form for error messages
refs #31454

GiudGiud committed:
Fix some mistakes in aqueous physics:
- addauxkernel not addauxvar
- size intermediate arrays
- fixup logic for darcy advection terms
- better error messages
- auxkernels take list of reactants not full list
refs #31454

GiudGiud committed:
Keep the parenthesis for the state
refs #31453

GiudGiud committed:
Add unit test for the reaction parser
refs #31453

GiudGiud committed:
Rename ReactionKineticsPhysicsBase to ReactionNetworkPhysicsBase because
Aqueous reactions equilibrium is actually not kinetics
refs #31454

GiudGiud committed:
Rename AqueousReactionKinetics to AqueousReactionsEquilibriumPhysics because
Aqueous reactions equilibrium is actually not kinetics
refs #31454

GiudGiud committed:
Add documentation for the reaction network physics
refs #31454

GiudGiud committed:
Add documentation for the new parsing utility
refs #31453

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3195731/), step Clang format on a23cd5e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31409/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31409/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ded93b763fb0759f8008c3b9b5992d9637ccdd6e`

GiudGiud committed:
Address test failures:
- missing header
- return false after mooseError: should not have been needed but mingcc requires it
Remove unused params

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3196277/), step Docs: sync website on 6419714 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31409/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3196313/), step Generate coverage on 6419714 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>0cc44f</th>
      <th colspan=4>#31409 641971</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>86.02%</td>
      <td>86.02%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>124557</td>
      <td>124557</td>
      <td>-</td>
      <td>9</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>20246</td>
      <td>20246</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31409/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31409/coverage/framework)


## Modules coverage

### Chemical reactions

<table>
  <thead>
    <tr>
      <th></th>
      <th>0cc44f</th>
      <th colspan=4>#31409 641971</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.63%</td>
      <td>91.73%</td>
      <td>-1.91%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1515</td>
      <td>1785</td>
      <td>+270</td>
      <td>328</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>103</td>
      <td>161</td>
      <td>+58</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31409/coverage_diff/chemical_reactions)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31409/coverage/chemical_reactions)

### Scalar transport

<table>
  <thead>
    <tr>
      <th></th>
      <th>0cc44f</th>
      <th colspan=4>#31409 641971</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>92.03%</td>
      <td>92.05%</td>
      <td>+0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>381</td>
      <td>382</td>
      <td>+1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>33</td>
      <td>33</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31409/coverage_diff/scalar_transport)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31409/coverage/scalar_transport)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31409/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3196313/), step Verify coverage on 6419714 wanted to post the following:


The following coverage requirement(s) failed:
- `chemical_reactions` coverage rate 91.73% is less than the required 92.0%

csdechant commented on modules/chemical_reactions/src/utils/ReactionNetworkUtils.C:
```diff
@@ -0,0 +1,308 @@
+  for (const auto & term : products)
+    species.push_back(term.species + term.state.value_or("") + term.charge.value_or(""));
+  return species;
+}
```
Clarifying question:
As rewritten, this allows for duplicate species names in a single reaction, correct? For example, someone inputs:
`e + Ar -> e + e + Ar+` for argon ionization, this would have three entries for electrons, `e` (one entry for reactions and two entries for products).

If so, that is good thing. I just wanted to point out that we might want to include a "consolidated function" here, or maybe just have the inheriting objects handle that?

csdechant replied:
Do you think it is worth having a "consolidated function" function here, to keep track of the total Stoichiometric coefficient, so should that be the job of the inheriting objects?

csdechant reviewed (commented):
Great work @GiudGiud!
I don't have any required changes, but I do have several questions.

GiudGiud commented on modules/chemical_reactions/doc/content/source/physics/AqueousReactionsEquilibriumPhysics.md:
```diff
@@ -0,0 +1,61 @@
+!alert note
+Only one product can be set per reaction and it should be an equilibrium species. Other products may
+be moved to the left hand side (LHS). If they are equilibrium species as well, they should set as the RHS in another
+equation.
```
This is a limitation of the downstream class (e.g. the code copy pasted from [AqueousEquilibriumReactions] and [SolidKineticReactions])
it s not a limitation of the reaction network parser. Once the reactions are parsed, the creation of kernels takes an extra step (based on a closure relation, consuming the metadata from the reaction), which requires more code.

The current code in that module has this limitation

>  has the traditional "reactions -> products" format in the text

exactly. it has it as you expect in the text, but then the code is not able to handle that!
Considering that code is custom for geochemical applications (despite the name of the module), I didnt want to work on improving that.

csdechant replied:
Cool, that sounds fair to me.

GiudGiud committed:
Add QA documentation to new tests

GiudGiud committed:
Add a set of 'unique' species and coefs getters to address review

GiudGiud committed:
Add unit testing for unique getters

GiudGiud committed:
Adapt convention for signs to what Corey mentioned

GiudGiud committed:
Add missing SQA documentation for tests folder modified

GiudGiud committed:
Update python tests due to resorting of an advanced parameter

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3546656/), step Results summary on a6616d4 wanted to post the following:

## Framework test summary

Compared against e4407b3 in job [civet.inl.gov/job/3542080](https://civet.inl.gov/job/3542080).

No change
## Modules test summary

Compared against e4407b3 in job [civet.inl.gov/job/3542080](https://civet.inl.gov/job/3542080).

### Added tests

| Test                                                                        | Time (s)   | Memory (MB)   |
|-----------------------------------------------------------------------------|------------|---------------|
| `chemical_reactions/test:aqueous_equilibrium.2species_with_density_physics` | 1.33       | 0.00          |
| `chemical_reactions/test:aqueous_equilibrium.1species_physics`              | 0.99       | 0.00          |
| `chemical_reactions/test:aqueous_equilibrium.co2_h2o_physics`               | 0.80       | 0.00          |

GiudGiud committed:
Add a test to increase coverage on variable creation
Fix logic around scaling factors

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3548235/), step Results summary on 4fab8c0 wanted to post the following:

## Framework test summary

Compared against e4407b3 in job [civet.inl.gov/job/3542080](https://civet.inl.gov/job/3542080).

No change
## Modules test summary

Compared against e4407b3 in job [civet.inl.gov/job/3542080](https://civet.inl.gov/job/3542080).

### Added tests

| Test                                                                        | Time (s)   | Memory (MB)   |
|-----------------------------------------------------------------------------|------------|---------------|
| `chemical_reactions/test:aqueous_equilibrium.2species_with_density_physics` | 1.41       | 0.00          |
| `chemical_reactions/test:aqueous_equilibrium.1species_physics`              | 0.99       | 0.00          |
| `chemical_reactions/test:aqueous_equilibrium.co2_h2o_physics`               | 0.77       | 0.00          |
| `chemical_reactions/test:aqueous_equilibrium.1species_physics_variables`    | 0.68       | 0.00          |

GiudGiud committed:
Further increase coverage:
- trigger verbose reaction output
- delete unused code
- use the option for specifying K over log_10K

GiudGiud commented:
Coverage failure should be gone now too. 92% is not that easy to achieve!

lindsayad reviewed (approved):
Looks like some good discussion with @csdechant. The class names look good to me

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3574722/), step Results summary on 25d94a9 wanted to post the following:

## Framework test summary

Compared against 019760d in job [civet.inl.gov/job/3572357](https://civet.inl.gov/job/3572357).

No change
## Modules test summary

Compared against 019760d in job [civet.inl.gov/job/3572357](https://civet.inl.gov/job/3572357).

### Added tests

| Test                                                                        | Time (s)   | Memory (MB)   |
|-----------------------------------------------------------------------------|------------|---------------|
| `chemical_reactions/test:aqueous_equilibrium.2species_with_density_physics` | 1.37       | 0.00          |
| `chemical_reactions/test:aqueous_equilibrium.1species_physics_variables`    | 1.08       | 0.00          |
| `chemical_reactions/test:aqueous_equilibrium.1species_physics`              | 0.95       | 0.00          |
| `chemical_reactions/test:aqueous_equilibrium.co2_h2o_physics`               | 0.73       | 0.00          |
